src/Libraries/Microsoft.Extensions.AI/ChatCompletion/FunctionInvokingChatClient.cs

```diff
@@ -562,13 +572,93 @@ FunctionResultContent CreateFunctionResultContent(FunctionInvocationResult resul
     /// </param>
     /// <param name="cancellationToken">The <see cref="CancellationToken"/> to monitor for cancellation requests. The default is <see cref="CancellationToken.None"/>.</param>
     /// <returns>The result of the function invocation. This may be null if the function invocation returned null.</returns>
-    protected virtual Task<object?> InvokeFunctionAsync(FunctionInvocationContext context, CancellationToken cancellationToken)
+    protected virtual async Task<object?> InvokeFunctionAsync(FunctionInvocationContext context, CancellationToken cancellationToken)
     {
         _ = Throw.IfNull(context);
 
-        return context.Function.InvokeAsync(context.CallContent.Arguments, cancellationToken);
+        using Activity? activity = _activitySource.StartActivity(context.Function.Metadata.Name);
+
+        long startingTimestamp = 0;
+        if (_logger.IsEnabled(LogLevel.Debug))
+        {
+            startingTimestamp = Stopwatch.GetTimestamp();
+            if (_logger.IsEnabled(LogLevel.Trace))
+            {
+                LogInvokingSensitive(context.Function.Metadata.Name, LoggingHelpers.AsJson(context.CallContent.Arguments));
```

Comment:
Shouldn't this be using the JSO from the corresponding AIFunction? This would invariably fail for nontrivial types in Native AOT.

Reply:
Alternatively, if serialization is sourced from `ToolCallJsonSerializationOptions` of the underlying chat service then perhaps there needs to be a way for this layer to see it.

Reply:
> Shouldn't this be using the JSO from the corresponding AIFunction? This would invariably fail for nontrivial types in Native AOT.

Yup, fixed.